backend/x11/backend.c

```diff
@@ -0,0 +1,374 @@
+#define _POSIX_C_SOURCE 199309L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <EGL/egl.h>
+#include <wayland-server.h>
+#include <xcb/xcb.h>
+#include <xcb/glx.h>
+#include <X11/Xlib-xcb.h>
+#include <linux/input-event-codes.h>
+#include <wlr/backend/interface.h>
+#include <wlr/backend/x11.h>
+#include <wlr/egl.h>
+#include <wlr/interfaces/wlr_output.h>
+#include <wlr/interfaces/wlr_input_device.h>
+#include <wlr/interfaces/wlr_keyboard.h>
+#include <wlr/interfaces/wlr_pointer.h>
+#include <wlr/util/log.h>
+#include "backend/x11.h"
+
+static struct wlr_backend_impl backend_impl;
+static struct wlr_input_device_impl input_impl;
+static struct wlr_keyboard_impl keyboard_impl;
+static struct wlr_pointer_impl pointer_impl;
+
+static uint32_t xcb_button_to_wl(uint32_t button) {
+	switch (button) {
+	case XCB_BUTTON_INDEX_1: return BTN_LEFT;
+	case XCB_BUTTON_INDEX_2: return BTN_MIDDLE;
+	case XCB_BUTTON_INDEX_3: return BTN_RIGHT;
+	// XXX: I'm not sure the scroll-wheel direction is right
+	case XCB_BUTTON_INDEX_4: return BTN_GEAR_UP;
+	case XCB_BUTTON_INDEX_5: return BTN_GEAR_DOWN;
+	default: return 0;
+	}
+}
+
+static bool handle_x11_event(struct wlr_x11_backend *x11, xcb_generic_event_t *event) {
+	struct wlr_x11_output *output = &x11->output;
+
+	switch (event->response_type) {
+	case XCB_EXPOSE: {
+		wl_signal_emit(&output->wlr_output.events.frame, output);
+		break;
```

Comment:
dead code?

Reply:
Wow, how did I not notice that?
I merged XCB_KEY_PRESS and XCB_KEY_RELEASE together and I guess I forgot to delete the old one.